x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/RemoteAccessAuthenticationService.java

```diff
@@ -0,0 +1,240 @@
+                // drop remote access authentication headers since we've read their values, and we want to maintain the invariant that
+                // either the remote access authentication header is in the context, or the authentication header, but not both
+                List.of(REMOTE_ACCESS_CLUSTER_CREDENTIAL_HEADER_KEY, REMOTE_ACCESS_AUTHENTICATION_HEADER_KEY)
+            )
+        ) {
+            final Supplier<ThreadContext.StoredContext> storedContextSupplier = threadContext.newRestorableContext(false);
+            authcContext.addAuthenticationToken(remoteAccessHeaders.clusterCredential());
+            authenticationService.authenticate(
+                authcContext,
+                new ContextPreservingActionListener<>(storedContextSupplier, ActionListener.wrap(authentication -> {
+                    assert authentication.isApiKey() : "initial authentication for remote access must be by API key";
+                    assert false == authentication.isRunAs() : "initial authentication for remote access cannot be run-as";
+                    final RemoteAccessAuthentication remoteAccessAuthentication = remoteAccessHeaders.remoteAccessAuthentication();
+                    validate(remoteAccessAuthentication);
+                    writeAuthToContext(
+                        authcContext,
+                        authentication.toRemoteAccess(maybeRewriteForSystemUser(remoteAccessAuthentication)),
+                        listener
+                    );
+                }, ex -> withRequestProcessingFailure(authcContext, ex, listener)))
+            );
+        }
+    }
+
+    private static RemoteAccessAuthentication maybeRewriteForSystemUser(RemoteAccessAuthentication remoteAccessAuthentication)
+        throws IOException {
+        final Subject receivedEffectiveSubject = remoteAccessAuthentication.getAuthentication().getEffectiveSubject();
+        final User user = receivedEffectiveSubject.getUser();
+        if (SystemUser.is(user)) {
+            return new RemoteAccessAuthentication(
+                Authentication.newInternalAuthentication(
+                    SystemUser.INSTANCE,
+                    receivedEffectiveSubject.getTransportVersion(),
+                    receivedEffectiveSubject.getRealm().getNodeName()
+                ),
+                new RoleDescriptorsIntersection(CROSS_CLUSTER_INTERNAL_ROLE)
+            );
+        } else if (User.isInternal(user)) {
+            throw new IllegalArgumentException(
+                "received cross cluster request from an unexpected internal user [" + user.principal() + "]"
+            );
+        } else {
+            return remoteAccessAuthentication;
+        }
+    }
+
+    public AuthenticationService getAuthenticationService() {
+        return authenticationService;
+    }
+
+    private record RemoteAccessHeaders(
+        ApiKeyService.ApiKeyCredentials clusterCredential,
+        RemoteAccessAuthentication remoteAccessAuthentication
+    ) {}
```

Comment:
It is worth to have this standalone with methods so it can be used by both QC (SecurityServerTransportInterceptor) and FC.

Reply:
👍 I'll do this in a follow up